File: elixir/markdown/markdown.exs
defmodule Markdown do
  @doc """
    Parses a given string with Markdown syntax and returns the associated HTML for that string.

    ## Examples

    iex> Markdown.parse("This is a paragraph")
    "<p>This is a paragraph</p>"

    iex> Markdown.parse("#Header!\n* __Bold Item__\n* _Italic Item_")
    "<h1>Header!</h1><ul><li><em>Bold Item</em></li><li><i>Italic Item</i></li></ul>"
  """
  @spec parse(String.t) :: String.t
  def parse(markdown) do
    markdown
    |> String.split("\n")
    |> Enum.map(&process(&1))
    |> Enum.join
    |> patch
  end

  defp process(text) do
    case text do
      "#" <> _ ->
        text
        |> parse_header_md_level
        |> enclose_with_header_tag
      "*" <> _ ->
        parse_list_md_level(text)
      _ ->
        text
        |> String.split
        |> enclose_with_paragraph_tag
    end
  end

  defp parse_header_md_level(header) do
    [h|t] = String.split(header)
    weight = String.length(h) |> to_string
    text = Enum.join(t, " ")

    {weight, text}
  end

  defp parse_list_md_level(level) do
    text =
      level
      |> String.trim_leading("* ")
      |> String.split
      |> join_words_with_tags

    "<li>" <> text <> "</li>"
  end

  defp enclose_with_header_tag({weight, text}) do
    "<h#{weight}>#{text}</h#{weight}>"
  end

  defp enclose_with_paragraph_tag(text) do
    "<p>#{join_words_with_tags(text)}</p>"
  end

  defp join_words_with_tags(text) do
    text
    |> Enum.map(&replace_md_with_tag(&1))
    |> Enum.join(" ")
  end

  defp replace_md_with_tag(word) do
    word
    |> replace_prefix_md
    |> replace_suffix_md
  end

  defp replace_prefix_md(word) do
    cond do
      word =~ ~r/^#{"__"}{1}/ -> String.replace(word, "__", "<strong>", global: false)
      word =~ ~r/^#{"_"}+/ -> String.replace(word, "_", "<em>", global: false)
      true -> word
    end
  end

  defp replace_suffix_md(word) do
    cond do
      word =~ ~r/#{"__"}/ -> String.replace(word, "__", "</strong>")
      word =~ ~r/#{"_"}/ -> String.replace(word, "_", "</em>")
      true -> word
    end
  end

  defp patch(list) do
    list
    |> String.replace("<li>", "<ul><li>", global: false)
    |> String.replace_suffix("</li>", "</li></ul>")
  end
end

File: python/markdown/markdown.py
import re


def parse_markdown(markdown):
    lines = markdown.split('\n')
    result = [parse_line(line) for line in lines]
    result = ''.join(result)
    unordered_list = re.findall(r'<li>.*</li>', result)

    if unordered_list:
        return result.replace(unordered_list[0],
                              wrap(unordered_list[0], 'ul'))
    return result


def parse_line(line):
    header = re.findall(r'^#+(?=\s)', line)
    italics = re.findall(r'\b_[a-zA-Z0-9\s]+_\b', line)
    bolds = re.findall(r'\b__[a-zA-Z0-9\s]+__\b', line)
    list_item = re.match(r'^\*.*', line)
    result = line

    for italic in italics:
        result = result.replace(italic, wrap(italic[1:-1], 'em'))

    for bold in bolds:
        result = result.replace(bold, wrap(bold[2:-2], 'strong'))

    if header:
        header = len(header[0])
        return wrap(result[header+1:], f'h{header}')

    if list_item:
        return wrap(result[2:], 'li')

    return wrap(result, 'p')


def wrap(string, tag):
    return f'<{tag}>' + string + f'</{tag}>'

